src/persist-txn/src/txn_cache.rs

```diff
@@ -263,11 +271,14 @@ impl<T: Timestamp + Lattice + TotalOrder + StepForward + Codec64> TxnsCacheState
         }
 
         // The most recent forget is set, which means it's not registered as of
-        // the latest information we have. Read to the current progress point
-        // normally.
+        // the latest information we have. Emit logical progress up to the
+        // forget and then read to the current progress point normally.
         let last_reg = data_times.last_reg();
-        if last_reg.forget_ts.is_some() {
-            if ts < self.progress_exclusive {
+        if let Some(forget_ts) = &last_reg.forget_ts {
+            assert!(*forget_ts < self.progress_exclusive);
+            if ts <= *forget_ts {
+                return EmitLogicalProgress(forget_ts.step_forward());
+            } else if ts < self.progress_exclusive {
                 return ReadDataTo(self.progress_exclusive.clone());
             } else {
                 // TODO(txn): Make sure to add a regression test for this when
```

Comment:
I think this TODO is done? Mind removing the else locally and checking the the unit test fails?

Reply:
The reason is that it falls through to the final 

```
        // Nope, all caught up, we have to wait.
        WaitForTxnsProgress
```

which returns the right thing.

Reply:
I'll remove the comment, but I actually like keeping the branch. It allows us to make the assumption below that the shard IS registered which simplifies the reasoning a bit.

Reply:
Yeah, happy keeping the logic, but in that case, it does seem like it'd be nice to have something in the test that fails without it

Reply:
I think it's not possible to have a test that fails without this branch since it is strictly redundant with the final return. However, I think it's actually fine to just delete this branch too since we have asserts below that check that the shard is currently registered. I just pushed a commit to delete this branch.